crates/common/tedge_config/src/sudo.rs

```diff
@@ -0,0 +1,53 @@
+use std::ffi::OsStr;
+use std::path::Path;
+use std::process::Command;
+use std::sync::Arc;
+
+use tracing::warn;
+
+use crate::TEdgeConfig;
+
+const SUDO: &str = "sudo";
+
+/// An object used to remember the user's sudo preference.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct SudoCommandBuilder {
+    path: Option<Arc<Path>>,
+    enabled: bool,
+}
+
+impl SudoCommandBuilder {
+    pub fn new(config: &TEdgeConfig) -> Self {
+        Self::enabled(config.sudo.enable)
+    }
+
+    pub fn enabled(enabled: bool) -> Self {
+        let path = match which::which(SUDO) {
+            Ok(sudo) => Some(Arc::from(sudo)),
+            Err(_) => None,
+        };
+
+        Self { path, enabled }
+    }
+
+    /// Checks if sudo is present in $PATH. If it's present and sudo is enabled,
+    /// prepare a [`Command`](std::process::Command) using sudo. Otherwise,
+    /// prepares a Command that starts `program` directly.
+    pub fn command<S: AsRef<OsStr>>(&self, program: S) -> Command {
+        if !self.enabled {
+            return Command::new(program);
+        }
+
+        match &self.path {
+            Some(sudo) => {
+                let mut c = Command::new(sudo.as_ref());
+                c.arg(program);
+                c
+            }
+            None => {
+                warn!("`sudo.enable` set to `true`, but sudo not found in $PATH");
```

Comment:
I would move this warning into `SudoCommandBuilder::enabled` to remove the `enabled` field, which is redundant.

Reply:
I think it would be easier for users if we emit any sudo-related warnings when the command actually gets created and the process is about to be spawned, though. One warning at the beginning is easy to miss, amongst all the other warnings that can appear when the agent is started.

And for that, we need to actually distinguish the case whether or not `sudo` is enabled by the user, but wasn't found in `$PATH`. If we wanted to save space, couldn't we just run `which` every time instead of saving the result? I'd think it's not that expensive and in exchange we'd eliminate the behaviour where currently, if `sudo` is in path when agent is started, but during runtime is removed (a bit of an edge case, I know), the agent would still try to run the old sudo path, which seems like an oversight.

For now I've kept the current behaviour because I wanted to avoid adding too much to the PR, but I think it would be an improvement.